sdks/python/container/boot.go

```diff
@@ -137,46 +145,49 @@ func main() {
 
 	options, err := provision.ProtoToJSON(info.GetPipelineOptions())
 	if err != nil {
-		log.Fatalf("Failed to convert pipeline options: %v", err)
+		return fmt.Errorf("Failed to convert pipeline options: %v", err)
 	}
 
 	// (2) Retrieve and install the staged packages.
 	//
-	// Guard from concurrent artifact retrieval and installation,
-	// when called by child processes in a worker pool.
+	// No log.Fatalf() from here on, otherwise deferred cleanups will not be called!
```

Comment:
Then why not simply put the parts that require defer cleanups in a function, instead of heavy handedly making all of main a new function with log fatal?

In this case, I'd put the latter half into a function "launchWorkerProcesses" which handles the defer required portions, with the artifact/provision contract portions handled in a "setupBeamRequirements" or similarly named function. This would make the "phases" obvious in main.

Reply:
As an aside, the boot.go mains are long overdue for refactoring to be more testable, and share code with each other. They all duplicate very common portions leading to repeated work in implementing features/capabilities in any of the pre-amble.

That's out of scope for this PR though.

Reply:
See above. log.Fatal() cancels all goroutines and therefore no cleanup will happen.

Reply:
But the part I was commenting on is before any defer calls are being made. This comment is about the factoring for making the function more readable and maintainable, improving the change being made, not reverting it.  

Basically the change is good, but the implementation scope is unnecessarily broad.  which puts additional strain on a reader for tracking which scope they're in, when instead this change can accomplish it's goal (ensuring the defered cleanups and goroutines are shutdown in an orderly fashion), and improve readability, via smaller, easier to understand functions composed with purpose.

Reply:
I also suggested splitting them into multiple functions, which would avoid that issue.

It's not prescribed in Go to "only have a single return" or "only have a single way to exit". You're also allowed to do so if you like.

Reply:
I moved some of the stuff at the beginning back to main, so it's basically two functions now. The rest either initialises variables that are needed later in the same scope or needs to return without exiting the main process.